cmd/cloud_sql_proxy/internal/healthcheck/healthcheck_test.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2021 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package healthcheck_test
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"testing"
+	"syscall"
+
+	"github.com/GoogleCloudPlatform/cloudsql-proxy/cmd/cloud_sql_proxy/internal/healthcheck"
+	"github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/proxy"
+)
+
+const (
+	livenessPath  = "/liveness"
+	readinessPath = "/readiness"
+	testPort      = "8090"
+)
+
+// Test to verify that when the proxy client is up, the liveness endpoint writes 200.
+func TestLiveness(t *testing.T) {
+	s, err := healthcheck.NewServer(&proxy.Client{}, testPort)
+	if err != nil {
+		t.Fatalf("Could not initialize health check: %v\n", err)
+	}
+	defer s.Close(context.Background())
+
+	resp, err := http.Get("http://localhost:" + testPort + livenessPath)
+	if err != nil {
+		t.Fatalf("HTTP GET failed: %v\n", err)
+	}
+	if resp.StatusCode != 200 {
+		t.Errorf("Got status code %v instead of 200\n", resp.StatusCode)
+	}
+}
+
+// Test to verify that 1. when startup has NOT finished, the readiness endpoint writes 500.
+// 2. when startup HAS finished (and MaxConnections limit not specified), the readiness
+// endpoint writes 200.
+func TestStartup(t *testing.T) {
+	cases := []struct {
+		finishedStartup bool
+		statusCode int
+	}{
+		{
+			finishedStartup: false,
+			statusCode: 500,
+		},
+		{
+			finishedStartup: true,
+			statusCode: 200,
+		},
+	}
+
+	for _, c := range cases {
+		func() {
+			s, err := healthcheck.NewServer(&proxy.Client{}, testPort)
```

Comment:
Do we need to start a server each time, or can this be done outside the loop?

Reply:
We don't have a function that flips `started` from `true` to `false`. It didn't seem like a great idea to rely on the order in which tests are performed (having to test the case when `started` is `false` before when `started` is `true`). 

It might be feel less "natural," but I could manually set `s.started = false` in the case where `finishedStartup` is false? This would remove the need to start a new server with each iteration of the loop.

Reply:
I agree we don't want to introduce a `s.started = false` clause. The anonymous function feels awkward here, but sticking with the table-based test doesn't make it easy to avoid. So I think we can leave this here for now and revisit it in future passes (refactor code and refactor tests, as they say). If it were me, I might just split out two separate tests (more duplication, but easier to understand), but I'll leave that decision up to you, as either approach is fine in the scheme of things.

Reply:
Decided to revert back to separate tests, but this was a very good learning experience for me.